### PR TITLE
Jkralik/feature/initializatio via user agent

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: service/device/config.go [hunk 1]
```diff
@@ -148,11 +148,47 @@ func (c *CoapConfig) Validate() error {
...
-	subjectUUID      uuid.UUID
-	PreSharedKeyUUID string `yaml:"preSharedKeyUuid" json:"preSharedKeyUuid"`
-	preSharedKeyUUID uuid.UUID
+	Authentication Authentication     `yaml:"authentication" json:"authentication"`
+	PreSharedKey   PreSharedKeyConfig `yaml:"preSharedKey" json:"preSharedKey"`
+}
+
+func (c *TLSConfig) Validate() error {
+	switch c.Authentication {
+	case "":
```
Comment (Danielius1922) [inline service/device/config.go:183]: Isn't any value besides `AuthenticationPreSharedKey` or `AuthenticationPreSharedKey` an error? Not just "".

Comment (jkralik) [reply service/device/config.go:183]: Good point :)

File: test/test.go [hunk 2]
```diff
@@ -248,9 +281,15 @@ func NewClientApplicationServer(ctx context.Context, opts ...ClientApplicationSe
 		defer wg.Done()
 		_ = d.Serve()
 	}()
+	remoteProvisioningCfg := cfg.RemoteProvisioningCfg
+	if err := remoteProvisioningCfg.Validate(); err != nil {
+		return nil, nil, err
```
Comment (Danielius1922) [inline test/test.go:286]: Probably 

```_ = d.Close()```

is missing here for corectness.

File: service/service.go [hunk 3]
```diff
@@ -20,75 +20,89 @@ import (
...
+		insecure = " (insecure)"
+	}
+	log.Infof("gRPC API available on %s%v", addr, insecure)
+	return grpcService, nil
+}
+
+func closeHttpOnError(err error, services []service.APIService) error {
+	if len(services) > 0 {
+		err = fmt.Errorf("cannot create grpc service: %w", err)
+		err2 := services[0].Close()
```
Comment (Danielius1922) [inline service/service.go:67]: Isn't `service[0]` always deviceService and not httpService (see line 86 below)?

File: service/remoteProvisioning/config.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	switch c.Mode {
+	case Mode_None:
+	case Mode_UserAgent:
+		if err := c.UserAgentConfig.Validate(); err != nil {
+			return fmt.Errorf("userAgent.%w", err)
+		}
+		if err := c.Authorization.Validate(c.Mode); err != nil {
+			return fmt.Errorf("authorization.%w", err)
+		}
+	}
```
Comment (Danielius1922) [inline service/remoteProvisioning/config.go:94]: default:
  return invalid value error

are unexpected values OK?

File: service/remoteProvisioning/config.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+		}
+		if c.Authority == "" {
+			return fmt.Errorf("authority('%v')", c.Authority)
+		}
+		if c.OwnerClaim == "" {
+			return fmt.Errorf("ownerClaim('%v')", c.OwnerClaim)
+		}
+	case Mode_None:
+	}
+	return nil
```
Comment (Danielius1922) [inline service/remoteProvisioning/config.go:75]: What happens when there is a value not equal to Mode_None or Mode_UserAgent ? I think it gets parsed and saved, so validation should return error here.

File: service/http/getJSONWebKeys_test.go [hunk 6]
```diff
@@ -17,24 +17,37 @@
...
 )
 
-func TestClientApplicationServerGetWebConfiguration(t *testing.T) {
-	cfg := test.MakeConfig(t)
-	cfg.APIs.HTTP.TLS.ClientCertificateRequired = false
-	cfg.APIs.HTTP.UI.Enabled = true
-	shutDown := test.New(t, cfg)
-	defer shutDown()
+func TestClientApplicationServerGetJSONWebKeys(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3600)
```
Comment (Danielius1922) [inline service/http/getJSONWebKeys_test.go:33]: 1 hour timeout seems a bit long for a test

Comment (jkralik) [reply service/http/getJSONWebKeys_test.go]: fixed

File: service/service.go (hunk 3 above)
Comment (jkralik) [reply service/service.go]: fixed

File: test/test.go (hunk 2 above)
Comment (jkralik) [reply test/test.go]: fixed

File: service/grpc/getJSONWebKeys.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	marshaledJwk, err := json.Marshal(map[string]interface{}{
+		"keys": keys,
+	})
+	if jwksCache == nil {
+		return nil, status.Errorf(codes.Internal, "cannot marshal keys to json: %v", err)
+	}
+	var jwkMap map[string]interface{}
+	err = json.Unmarshal(marshaledJwk, &jwkMap)
+	if jwksCache == nil {
```
Comment (Danielius1922) [inline service/grpc/getJSONWebKeys.go:54]: Wrong check I think, should be

```err != nil```

`jwksCache == nil` is checked already on line 49

File: service/grpc/jsonWebKeyCache.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	return nil, fmt.Errorf("could not find JWK")
+}
+
+func (s *ClientApplicationServer) ParseWithClaims(token string, claims jwt.Claims) error {
+	c := s.jwksCache.Load()
+	if c == nil {
+		return status.Errorf(codes.Unauthenticated, "cannot validate token: missing JWK cache")
+	}
+	if token == "" {
```
Comment (Danielius1922) [inline service/grpc/jsonWebKeyCache.go:59]: Can be moved up to line 55, so the atomic load of s.jwksCache is not performed if the token is empty.

File: service/grpc/server.go [hunk 9]
```diff
@@ -18,53 +18,69 @@ package grpc
...
 }
 
 func (s *ClientApplicationServer) Version() string {
 	return s.info.Version
 }
 
-func (s *ClientApplicationServer) getDevice(deviceID string) (*device, error) {
-	d, ok := s.devices.Load(deviceID)
+func (s *ClientApplicationServer) Close() {
+	s.csrCache.Stop()
```
Comment (Danielius1922) [inline service/grpc/server.go:68]: Doesn't s.remoteOwnSignCache need to be stopped as well?

File: service/grpc/server.go [hunk 10]
```diff
@@ -18,53 +18,69 @@ package grpc
...
 
-func (s *ClientApplicationServer) deleteDevice(ctx context.Context, deviceID string) error {
-	d, ok := s.devices.LoadAndDelete(deviceID)
+func (s *ClientApplicationServer) deleteDevice(ctx context.Context, deviceID uuid.UUID) error {
+	dev, ok := s.devices.PullOut(deviceID)
 	if !ok {
 		return nil
 	}
-	dev, ok := d.(*device)
 	if !ok {
```
Comment (Danielius1922) [inline service/grpc/server.go:84]: Duplicit check with the one on line 81

File: service/http/service.go [hunk 11]
```diff
@@ -144,11 +175,15 @@ func New(ctx context.Context, serviceName string, config Config, clientApplicati
 
 // Serve starts the service's HTTP server and blocks
 func (s *Service) Serve() error {
-	return s.httpServer.Serve(s.listener)
+	err := s.httpServer.Serve(s.listener)
+	if errors.Is(err, http.ErrServerClosed) {
+		return nil
+	}
+	return nil
```
Comment (Danielius1922) [inline service/http/service.go:182]: This returns nil in both branches, shouldn't the final line be

`return err`

?

File: service/grpc/ownDevice.go [hunk 12]
```diff
@@ -19,21 +19,185 @@ package grpc
...
+}
+
+func (s *remoteSign) SendCertificate(ctx context.Context, certificate []byte) error {
+	select {
+	case err := <-s.errChan:
+		return err
+	case <-ctx.Done():
+		return fmt.Errorf("cannot send certificate: %w", ctx.Err())
+	case <-s.ctx.Done():
+		return fmt.Errorf("cannot send certificate: %w", ctx.Err())
```
Comment (Danielius1922) [inline service/grpc/ownDevice.go:86]: `ctx.Err()` or `s.ctx.Err()` ?

File: service/grpc/server.go (hunk 9 above)
Comment (jkralik) [reply service/grpc/server.go:68]: no it is `sync.Map` type

File: service/grpc/jsonWebKeyCache.go (hunk 8 above)
Comment (jkralik) [reply service/grpc/jsonWebKeyCache.go:59]: fixed

File: service/grpc/getJSONWebKeys.go (hunk 7 above)
Comment (jkralik) [reply service/grpc/getJSONWebKeys.go:54]: fixed

File: service/grpc/ownDevice.go (hunk 12 above)
Comment (jkralik) [reply service/grpc/ownDevice.go:86]: fixed

File: service/grpc/server.go (hunk 10 above)
Comment (jkralik) [reply service/grpc/server.go:84]: removed

**Review by ondrejtomcik — APPROVED** [review]

**Pull request merged**